src/python_testing/TC_DEM_2_4.py

```diff
@@ -0,0 +1,367 @@
+            TestStep("18a", "TH reads ESAState.",
+                     "Verify value is 0x01 (Online)"),
+            TestStep("19", "TH sends ResumeRequest.",
+                     "Verify DUT responds with status INVALID_IN_STATE(0xcb)"),
+            TestStep("20", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TEST_EVENT_TRIGGER_KEY and EventTrigger field set to PIXIT.DEM.TEST_EVENT_TRIGGER for Pausable Test Event Clear.",
+                     "Verify DUT responds with status SUCCESS(0x00)"),
+        ]
+
+        return steps
+
+    @async_test_body
+    async def test_TC_DEM_2_4(self):
+
+        logging.info(Clusters.Objects.DeviceEnergyManagement.Attributes.FeatureMap)
+
+        self.step("1")
+        # Commission DUT - already done
+
+        # Subscribe to Events and when they are sent push them to a queue for checking later
+        events_callback = EventChangeCallback(Clusters.DeviceEnergyManagement)
+        await events_callback.start(self.default_controller,
+                                    self.dut_node_id,
+                                    self.matter_test_config.endpoint)
+
+        self.step("2")
+        await self.check_test_event_triggers_enabled()
+
+        self.step("3")
+        await self.send_test_event_trigger_pausable()
+
+        self.step("3a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("3b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+
+        asserts.assert_not_equal(forecast, NullValue)
+        asserts.assert_equal(forecast.isPausable, True)
+        asserts.assert_greater(forecast.slots[0].minPauseDuration, 1)
+        asserts.assert_greater(forecast.slots[0].maxPauseDuration, 1)
+        asserts.assert_equal(forecast.slots[0].slotIsPausable, True)
+        asserts.assert_equal(forecast.slots[1].slotIsPausable, False)
+        asserts.assert_equal(forecast.activeSlotNumber, 0)
+
+        if forecast is not NullValue:
+            asserts.assert_less_equal(forecast.earliestStartTime, forecast.startTime,
+                                      f"Expected forecast earliestStartTime {forecast.earliestStartTime} to be <= startTime {forecast.startTime}")
+            asserts.assert_greater_equal(forecast.latestEndTime, forecast.endTime,
+                                         f"Expected forecast latestEndTime {forecast.latestEndTime} to be >= endTime {forecast.endTime}")
+            asserts.assert_equal(forecast.forecastUpdateReason, Clusters.DeviceEnergyManagement.Enums.ForecastUpdateReasonEnum.kInternalOptimization,
+                                 f"Expected forecast forecastUpdateReason {forecast.forecastUpdateReason} to be == Clusters.DeviceEnergyManagement.Enums.ForecastUpdateReasonEnum.kInternalOptimization")
+            for index, slot in enumerate(forecast.slots):
```

Comment:
We could move this forecast print into a function (maybe in DEMTestBase) if it is used in other TCs.

Reply:
Done